torch/csrc/dynamo/guards.cpp

```diff
@@ -574,6 +574,480 @@ static struct PyModuleDef _module = {
+      : GuardAccessor(name), _attr_name(name.ptr()) {}
+
+  PyObject* access(PyObject* obj) override { // borrowed ref
+    PyObject* ret = PyObject_GetItem(obj, _attr_name);
+    // Store the owning reference to _accessed_value to be decref'd on
+    // destruction.
+    _accessed_value = ret;
+    return ret;
+  }
+
+ private:
+  PyObject* _attr_name;
+};
+
+/**
+ * GuardManager encapsulates all the guards related to a particular py::object.
+ * It is a tree structure and consists of
+ * 1) Leaf guards - Guards that are run on the user given object
+ * 2) Accessors - Guard accessors (like getattr, getitem) to access the next
+ * value in the tree hierarchy. Accessor object also holds the child
+ * GuardManager.
+ *
+ * Lets look at an example to understand how it works.
+ * class Pair:
+ *     int x = 1;
+ *     int y = 2;
+ *
+ * At compile time
+ * >> guard_mananger = GuardManager()
+ * >> guard_mananger.x.add_lambda_guard(
+ *        lambda x: isinstance(x, Pair),
+ *        lambda x: f"expected Pair, found {type(x)}"
+ *    )
+ * >> guard_mananger.x.add_lambda_guard(lambda x: x == 1, lambda x: f"found {x},
+ * expected 1")
+ * >> guard_mananger.y.add_lambda_guard(lambda x: x == 2, lambda x: f"found {x},
+ * expected 2")
+ *
+ * At runtime
+ * >> guard_mananger.check(Pair())
+ *
+ * At compile time we build the tree structure. When we do `guard_manager.x`, it
+ * creates an AttrGuardAccessorNode, initializes a child guard manager with this
+ * accessor node, and adds it as a child. When we do
+ * `guard_manager.x.add_lambda_guard`, we call add_lambda_guard on the newly
+ * created guard manager and register a new leaf guard on it.
+ *
+ * At runtime, the accessor node has an important function of providing a way to
+ * access the value for the child guard. In the above example, guard_manager.x
+ * adds an AttrGuardAccessorNode with attr_name x. When check function is
+ * called, parent GuardManager calls getattr(value, "x") on its value passed to
+ * the check function to call the check function of the child guard manager.
+ *
+ * Performace optimization for fail fast - An optimization for runtime here is
+ * to sort the execution of child guards depending on the failure count.  This
+ * ensures that we run the guards that are more prone to fail statistically
+ * first. This can improve the cache lookup time when we have multiple cache
+ * entries.
+ */
+class GuardManager {
+ public:
+  GuardManager() = default;
+  GuardManager(const GuardManager& m) = delete;
+  GuardManager& operator=(const GuardManager&) = delete;
+
+  // GuardManager is the owner of the leaf_guards
+  void add_leaf_guard(std::shared_ptr<LeafGuard> leaf_guard) {
+    _leaf_guards.push_back(leaf_guard);
+  }
+
+  /**
+   * Adds a new guard manager with appropriate Accessor. If the accessor is
+   * already present, we just return the guard manager.
+   */
+  template <typename GuardAccessorT>
+  GuardManager* get_child_manager(py::object accessor_key) {
```

Comment:
```suggestion
  GuardManager* get_child_manager(const py::object& accessor_key) {
```
incref seems unnessary.